backend/src/Designer/Services/Implementation/ModelNameValidator.cs

```diff
@@ -0,0 +1,111 @@
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Text.Unicode;
+using System.Threading.Tasks;
+using System.Xml;
+using System.Xml.Schema;
+using Altinn.App.Core.Models;
+using Altinn.Studio.DataModeling.Converter.Interfaces;
+using Altinn.Studio.DataModeling.Converter.Json.Strategy;
+using Altinn.Studio.DataModeling.Converter.Metadata;
+using Altinn.Studio.DataModeling.Metamodel;
+using Altinn.Studio.Designer.Enums;
+using Altinn.Studio.Designer.Exceptions.DataModeling;
+using Altinn.Studio.Designer.Models;
+using Altinn.Studio.Designer.Services.Interfaces;
+using Json.Schema;
+
+namespace Altinn.Studio.Designer.Services.Implementation;
+
+public class ModelNameValidator : IModelNameValidator
+{
+    private readonly IXmlSchemaToJsonSchemaConverter _xmlSchemaToJsonSchemaConverter;
+    private readonly IJsonSchemaToXmlSchemaConverter _jsonSchemaToXmlSchemaConverter;
+    private readonly IModelMetadataToCsharpConverter _modelMetadataToCsharpConverter;
+    private readonly IAltinnGitRepositoryFactory _altinnGitRepositoryFactory;
+
+    public ModelNameValidator(IXmlSchemaToJsonSchemaConverter xmlSchemaToJsonSchemaConverter, IJsonSchemaToXmlSchemaConverter jsonSchemaToXmlSchemaConverter, IModelMetadataToCsharpConverter modelMetadataToCsharpConverter, IAltinnGitRepositoryFactory altinnGitRepositoryFactory)
+    {
+        _xmlSchemaToJsonSchemaConverter = xmlSchemaToJsonSchemaConverter;
+        _jsonSchemaToXmlSchemaConverter = jsonSchemaToXmlSchemaConverter;
+        _modelMetadataToCsharpConverter = modelMetadataToCsharpConverter;
+        _altinnGitRepositoryFactory = altinnGitRepositoryFactory;
+    }
+
+
+    public async Task ValidateModelNameForNewXsdSchemaAsync(Stream xsdSchema, string fileName, AltinnRepoEditingContext altinnRepoEditingContext)
+    {
+        var repository = _altinnGitRepositoryFactory.GetAltinnAppGitRepository(altinnRepoEditingContext.Org, altinnRepoEditingContext.Repo, altinnRepoEditingContext.Developer);
+
+        AltinnRepositoryType altinnRepositoryType = await repository.GetRepositoryType();
+        if (altinnRepositoryType != AltinnRepositoryType.App)
+        {
+            return;
+        }
+
+        // Try to go through conversion all the way.
+        // This will enforce either all files are written to disk or none if there is an error.
+        using XmlReader xmlReader = XmlReader.Create(xsdSchema);
+        var xmlSchema = XmlSchema.Read(xmlReader, (_, _) => { });
+        xsdSchema.Seek(0, SeekOrigin.Begin);
+        var jsonSchema = _xmlSchemaToJsonSchemaConverter.Convert(xmlSchema);
+        var jsonSchemaConverterStrategy = JsonSchemaConverterStrategyFactory.SelectStrategy(jsonSchema);
+        var metamodelConverter = new JsonSchemaToMetamodelConverter(jsonSchemaConverterStrategy.GetAnalyzer());
+        var modelMetadata = metamodelConverter.Convert(SerializeJsonSchema(jsonSchema));
+        _modelMetadataToCsharpConverter.CreateModelFromMetadata(modelMetadata);
```

Comment:
Could this be separated to a method so it is more clear where we test the conversion before writing?

Reply:
Extracted to method now.